include/Sema/Constraints.hpp

```diff
@@ -0,0 +1,466 @@
+
+    ///
+    /// @brief Constructs a constraint with a conversion restriction.
+    ///
+    /// @param kind Constraint kind.
+    /// @param restriction Conversion restriction kind.
+    /// @param first Source type.
+    /// @param second Target type.
+    /// @param locator AST source node.
+    /// @param typeVars Type variables involved.
+    ///
+    Constraint(
+        ConstraintKind kind, ConversionRestrictionKind restriction,
+        glu::types::Ty first, glu::types::Ty second, glu::ast::ASTNode *locator,
+        llvm::SmallPtrSetImpl<glu::types::TypeVariableTy *> &typeVars
+    );
+
+    /// @brief Constructs a constraint with a Member.
+    /// @param kind Constraint kind.
+    /// @param first First type.
+    /// @param second Second type.
+    /// @param member The member expression.
+    /// @param locator AST source node.
+    /// @param typeVars Type variables involved.
+    Constraint(
+        ConstraintKind kind, glu::types::Ty first, glu::types::Ty second,
+        glu::ast::StructMemberExpr *member, glu::ast::ASTNode *locator,
+        llvm::SmallPtrSetImpl<glu::types::TypeVariableTy *> &typeVars
+    );
+
+    /// @brief Constructs a syntactic element constraint.
+    ///
+    /// @param node The AST node representing the syntactic element.
+    /// @param isDiscarded Whether the result of this constraint is unused.
+    /// @param locator The AST node that triggered this constraint.
+    /// @param typeVars Type variables involved in the constraint.
+    ///
+    Constraint(
+        glu::ast::ASTNode node, bool isDiscarded, glu::ast::ASTNode *locator,
+        llvm::SmallPtrSetImpl<glu::types::TypeVariableTy *> &typeVars
+    );
+
+    /// @brief Constructs a bind overload constraint.
+    ///
+    /// @param type The type to be constrained.
+    /// @param choice The function declaration representing the overload choice.
+    /// @param locator The AST node that triggered this constraint.
+    /// @param typeVars Type variables involved in the constraint.
+    ///
+    Constraint(
+        glu::types::Ty type, glu::ast::FunctionDecl *choice,
+        glu::ast::ASTNode *locator,
+        llvm::SmallPtrSetImpl<glu::types::TypeVariableTy *> &typeVars
+    );
+
+    ///
+    /// @brief Gets the mutable buffer of type variables.
+    ///
+    /// @return Mutable array of type variables.
+    ///
+    llvm::MutableArrayRef<glu::types::TypeVariableTy *> getTypeVariablesBuffer()
+    {
+        return { getTrailingObjects<glu::types::TypeVariableTy *>(),
+                 _numTypeVariables };
+    }
+
+    template <typename T>
+    size_t numTrailingObjects(TrailingObjects::OverloadToken<T>) const
+    {
+        if constexpr (std::is_same_v<T, glu::types::TypeVariableTy *>) {
+            return _numTypeVariables;
+        } else if constexpr (std::is_same_v<T, glu::ast::FunctionDecl *>) {
+            return (_kind == ConstraintKind::BindOverload) ? 1 : 0;
+        } else {
+            return 0;
+        }
+    }
```

Comment:
you're supposed to write different function overloads... not a templated one with an if?